File: backend/internal/hos/violations.go
package hos

import "time"

// violationOrder keeps the output deterministic across implementations.
var violationOrder = []string{
	ViolationDriveLimit, ViolationShiftLimit, ViolationBreakRequired, ViolationCycleLimit,
}

// collector records at most one entry per type per log day; a violation always
// supersedes the warning of the same type (Q57).
type collector struct {
	p        Policy
	dayStart time.Time
	dayEnd   time.Time
	found    map[string]Violation
}

func newCollector(p Policy, dayStart, dayEnd time.Time) *collector {
	return &collector{p: p, dayStart: dayStart, dayEnd: dayEnd, found: map[string]Violation{}}
}

func (c *collector) add(typ, severity string, at time.Time) {
	if at.Before(c.dayStart) || !at.Before(c.dayEnd) {
		return
	}
	prev, ok := c.found[typ]
	if ok && (prev.Severity == SeverityViolation || severity == prev.Severity) {
		return
	}
	c.found[typ] = Violation{Type: typ, Severity: severity, At: at.UTC()}
}

func (c *collector) result() []Violation {
	out := make([]Violation, 0, len(c.found))
	for _, typ := range violationOrder {
		if v, ok := c.found[typ]; ok {
			out = append(out, v)
		}
	}
	return out
}

// exceedAt returns the instant an accumulator that starts at acc and grows for d
// passes strictly beyond target inside the segment starting at start.
func exceedAt(acc, d time.Duration, start time.Time, target time.Duration) (time.Time, bool) {
	if target < 0 {
		target = 0
	}
	if acc+d <= target {
		return time.Time{}, false
	}
	if acc >= target {
		return start, true
	}
	return start.Add(target - acc), true
}

// reachAt returns the instant the accumulator reaches target (non strict).
func reachAt(acc, d time.Duration, start time.Time, target time.Duration) (time.Time, bool) {
	if target < 0 {
		target = 0
	}
	if acc+d < target {
		return time.Time{}, false
	}
	if acc >= target {
		return start, true
	}
	return start.Add(target - acc), true
}

// track records the warning/violation crossings of one accumulator counter.
func (c *collector) track(typ string, acc, d time.Duration, start time.Time, limit, warn time.Duration, canViolate bool) {
	if canViolate {
		if at, ok := exceedAt(acc, d, start, limit); ok {
			c.add(typ, SeverityViolation, at)
		}
	}
	if at, ok := reachAt(acc, d, start, limit-warn); ok {
		c.add(typ, SeverityWarning, at)
	}
}

// observe inspects one segment before it is applied to the state.
func (c *collector) observe(st *state, sg segment) {
	if !sg.isDuty() {
		return
	}
	d := sg.dur()
	driving := sg.status == StatusDrive
	th := c.p.WarningThresholds

	if driving {
		// Q10.4 + drive_limit: DR beyond 11h / DR beyond 8h without a break.
		c.track(ViolationDriveLimit, st.driveSinceRest, d, sg.start,
			c.p.driveLimit(), minutes(th.DriveMin), true)
		c.track(ViolationBreakRequired, st.driveSinceBreak, d, sg.start,
			c.p.breakAfter(), minutes(th.BreakMin), true)
	}
	// Q10.5: cycle grows on ON and DR, but only driving over the cycle is a violation.
	if cs, ce, ok := overlapRange(sg.start, sg.end, st.cycleFrom); ok {
		c.track(ViolationCycleLimit, st.cycleUsed, ce.Sub(cs), cs,
			c.p.cycleLimit(), minutes(th.CycleMin), driving)
	}
	// Q10.3: the 14h window is wall clock from the shift start; the segment that
	// opens the shift is observed before the state knows about it.
	shiftStart, paused := sg.start, time.Duration(0)
	if st.shiftActive {
		shiftStart, paused = st.shiftStart, st.shiftPaused
	}
	windowEnd := shiftStart.Add(c.p.shiftWindow() + paused)
	if driving && sg.end.After(windowEnd) {
		c.add(ViolationShiftLimit, SeverityViolation, laterOf(sg.start, windowEnd))
	}
	warnAt := windowEnd.Add(-minutes(th.ShiftMin))
	if !sg.end.Before(warnAt) {
		c.add(ViolationShiftLimit, SeverityWarning, laterOf(sg.start, warnAt))
	}
}

func laterOf(a, b time.Time) time.Time {
	if a.After(b) {
		return a
	}
	return b
}

// overlapRange clips [start,end) to the part at or after from.
func overlapRange(start, end, from time.Time) (time.Time, time.Time, bool) {
	if start.Before(from) {
		start = from
	}
	if !end.After(start) {
		return start, end, false
	}
	return start, end, true
}

// Violations evaluates one log day in the home terminal timezone (Q10.2, Q57).
// The whole history is replayed so that counters carry over from previous days;
// a status left open at the end of the day is projected to midnight.
func Violations(events []Event, p Policy, day time.Time, loc *time.Location) []Violation {
	p = p.normalized()
	if loc == nil {
		loc = time.UTC
	}
	evs, err := sortedEvents(events, p)
	if err != nil {
		return nil
	}
	dayStart, dayEnd := DayRange(day, loc)
	c := newCollector(p, dayStart.UTC(), dayEnd.UTC())
	st := newState(p, CycleWindowStart(dayStart, p, loc))
	st.obs = c.observe
	for _, sg := range walkSegments(evs, dayEnd.UTC()) {
		st.advance(sg)
	}
	return c.result()
}

// FormManner reports the form & manner state of a log day (Q57): missing
// trailer or shipping document is a warning while the day is open and becomes a
// violation once the day has been certified.
func FormManner(hasTrailer, hasDoc, certified bool) []Violation {
	severity := SeverityWarning
	if certified {
		severity = SeverityViolation
	}
	out := make([]Violation, 0, 2)
	if !hasTrailer {
		out = append(out, Violation{Type: ViolationFormMannerTrailer, Severity: severity})
	}
	if !hasDoc {
		out = append(out, Violation{Type: ViolationFormMannerDoc, Severity: severity})
	}
	return out
}

File: mobile/packages/hos_engine/lib/src/violations.dart
// Port of backend/internal/hos/violations.go (Q10.3, Q10.4, Q10.5, Q57).
import 'package:timezone/timezone.dart' as tz;

import 'compute.dart';
import 'day.dart';
import 'model.dart';
import 'policy.dart';

/// Go: `violationOrder` — keeps the output deterministic across
/// implementations.
const List<String> violationOrder = <String>[
  ViolationType.driveLimit,
  ViolationType.shiftLimit,
  ViolationType.breakRequired,
  ViolationType.cycleLimit,
];

/// Go: `collector` — records at most one entry per type per log day; a
/// violation always supersedes the warning of the same type (Q57).
class _Collector {
  _Collector(this.p, this.dayStart, this.dayEnd);

  final HosPolicy p;
  final DateTime dayStart;
  final DateTime dayEnd;
  final Map<String, HosViolation> found = <String, HosViolation>{};

  void add(String type, String severity, DateTime at) {
    if (at.isBefore(dayStart) || !at.isBefore(dayEnd)) return;
    final prev = found[type];
    if (prev != null && (prev.severity == Severity.violation || severity == prev.severity)) {
      return;
    }
    found[type] = HosViolation(type: type, severity: severity, occurredAt: utcOf(at));
  }

  List<HosViolation> result() {
    final out = <HosViolation>[];
    for (final type in violationOrder) {
      final v = found[type];
      if (v != null) out.add(v);
    }
    return out;
  }

  /// Go: `collector.track` — records the warning/violation crossings of one
  /// accumulator counter.
  void track(
    String type,
    Duration acc,
    Duration d,
    DateTime start,
    Duration limit,
    Duration warn, {
    required bool canViolate,
  }) {
    if (canViolate) {
      final at = exceedAt(acc, d, start, limit);
      if (at != null) add(type, Severity.violation, at);
    }
    final at = reachAt(acc, d, start, limit - warn);
    if (at != null) add(type, Severity.warning, at);
  }

  /// Go: `collector.observe` — inspects one segment before it is applied.
  void observe(HosState st, Segment sg) {
    if (!sg.isDuty) return;
    final d = sg.dur;
    final driving = sg.status == DutyStatus.dr;
    final th = p.warningThresholds;

    if (driving) {
      // Q10.4 + drive_limit: DR beyond 11h / DR beyond 8h without a break.
      track(
        ViolationType.driveLimit,
        st.driveSinceRest,
        d,
        sg.start,
        p.driveLimit,
        Duration(minutes: th.driveMin),
        canViolate: true,
      );
      track(
        ViolationType.breakRequired,
        st.driveSinceBreak,
        d,
        sg.start,
        p.breakAfter,
        Duration(minutes: th.breakMin),
        canViolate: true,
      );
    }
    // Q10.5: the cycle grows on ON and DR, but only driving over the cycle is
    // a violation.
    final clipped = overlapRange(sg.start, sg.end, st.cycleFrom);
    if (clipped != null) {
      track(
        ViolationType.cycleLimit,
        st.cycleUsed,
        clipped.end.difference(clipped.start),
        clipped.start,
        p.cycleLimit,
        Duration(minutes: th.cycleMin),
        canViolate: driving,
      );
    }
    // Q10.3: the 14h window is wall clock from the shift start; the segment
    // that opens the shift is observed before the state knows about it.
    var shiftStart = sg.start;
    var paused = Duration.zero;
    if (st.shiftActive) {
      shiftStart = st.shiftStart!;
      paused = st.shiftPaused;
    }
    final windowEnd = shiftStart.add(p.shiftWindow + paused);
    if (driving && sg.end.isAfter(windowEnd)) {
      add(ViolationType.shiftLimit, Severity.violation, laterOf(sg.start, windowEnd));
    }
    final warnAt = windowEnd.subtract(Duration(minutes: th.shiftMin));
    if (!sg.end.isBefore(warnAt)) {
      add(ViolationType.shiftLimit, Severity.warning, laterOf(sg.start, warnAt));
    }
  }
}

/// Go: `exceedAt` — the instant an accumulator that starts at [acc] and grows
/// for [d] passes strictly beyond [target] inside the segment starting at
/// [start]. Null when it never does.
DateTime? exceedAt(Duration acc, Duration d, DateTime start, Duration target) {
  var t = target;
  if (t < Duration.zero) t = Duration.zero;
  if (acc + d <= t) return null;
  if (acc >= t) return start;
  return start.add(t - acc);
}

/// Go: `reachAt` — the instant the accumulator reaches [target] (non strict).
DateTime? reachAt(Duration acc, Duration d, DateTime start, Duration target) {
  var t = target;
  if (t < Duration.zero) t = Duration.zero;
  if (acc + d < t) return null;
  if (acc >= t) return start;
  return start.add(t - acc);
}

DateTime laterOf(DateTime a, DateTime b) => a.isAfter(b) ? a : b;

/// Go: `overlapRange` — clips `[start, end)` to the part at or after [from].
({DateTime start, DateTime end})? overlapRange(DateTime start, DateTime end, DateTime from) {
  var s = start;
  if (s.isBefore(from)) s = from;
  if (!end.isAfter(s)) return null;
  return (start: s, end: end);
}

/// Go: `Violations` — evaluates one log day in the home terminal timezone
/// (Q10.2, Q57).
///
/// The whole history is replayed so that counters carry over from previous
/// days; a status left open at the end of the day is projected to midnight.
List<HosViolation> violations(
  List<HosEvent> events,
  HosPolicy policy,
  DateTime day,
  tz.Location location,
) {
  final p = policy.normalized();
  final evs = sortedEvents(events, p);
  final range = dayRange(day, location);
  final dayStart = utcOf(range.start);
  final dayEnd = utcOf(range.end);
  final c = _Collector(p, dayStart, dayEnd);
  final st = HosState(p, utcOf(cycleWindowStart(range.start, p, location)));
  st.obs = c.observe;
  for (final sg in walkSegments(evs, dayEnd)) {
    st.advance(sg);
  }
  return c.result();
}

/// Go: `FormManner` — the form & manner state of a log day (Q57): a missing
/// trailer or shipping document is a warning while the day is open and becomes
/// a violation once the day has been certified.
List<HosViolation> formManner({
  required bool hasTrailer,
  required bool hasDoc,
  required bool certified,
}) {
  final severity = certified ? Severity.violation : Severity.warning;
  final out = <HosViolation>[];
  if (!hasTrailer) {
    out.add(HosViolation(type: ViolationType.formMannerTrailer, severity: severity));
  }
  if (!hasDoc) {
    out.add(HosViolation(type: ViolationType.formMannerDoc, severity: severity));
  }
  return out;
}
